File: solutions/cpp/complex-numbers/1/complex_numbers.cpp
#include "complex_numbers.h"

namespace complex_numbers {
    Complex::Complex(double real_new, double imag_new) {
        real_ = real_new;
        imag_ = imag_new;
    }
    double Complex::real() const {
        return real_;
    }
    double Complex::imag() const {
        return imag_;
    }
    double Complex::abs() const {
        return sqrt(pow(real_, 2) + pow(imag_, 2));
    }
    Complex Complex::conj() const {
        return Complex(real_, imag_ * -1);
    }
    Complex Complex::exp() const {
        return pow(std::exp(1.0) , real_) * Complex(cos(imag_), sin(imag_));
    }

    // rhs and complex operations
    Complex Complex::operator+ (const Complex& rhs) const {
        return Complex(real_ + rhs.real_, imag_ + rhs.imag_);
    }
    Complex Complex::operator- (const Complex& rhs) const {
        return Complex(real_ - rhs.real_, imag_ - rhs.imag_);
    }
    Complex Complex::operator* (const Complex& rhs) const {
        return Complex(real_ * rhs.real_ - imag_ * rhs.imag_, imag_ * rhs.real_ + real_ * rhs.imag_);
    }
    Complex Complex::operator/ (const Complex& rhs) const {
        return Complex((real_ * rhs.real_ + imag_ * rhs.imag_) / (pow(rhs.real_, 2) + pow(rhs.imag_, 2)) , (imag_ * rhs.real_ - real_ * rhs.imag_) / (pow(rhs.real_, 2) + pow(rhs.imag_, 2)));
    }
    Complex Complex::operator+ (const double& rhs) const {
        return Complex(real_ + rhs, imag_);
    }
    Complex Complex::operator- (const double& rhs) const {
        return Complex(real_ - rhs, imag_);
    }
    Complex Complex::operator* (const double& rhs) const {
        return Complex(real_ * rhs, imag_ * rhs);
    }
    Complex Complex::operator/ (const double& rhs) const {
        return Complex(real_ / rhs, imag_ / rhs);
    }

    // lhs and complex operations
    Complex operator+ (double lhs, const Complex& rhs) {
        return Complex(lhs + rhs.real_, rhs.imag_);
    }
    Complex operator- (double lhs, const Complex& rhs) {
        return Complex(lhs - rhs.real_, 0 - rhs.imag_);
    }
    Complex operator* (double lhs, const Complex& rhs) {
        return Complex(lhs * rhs.real_, lhs * rhs.imag_);
    }
    Complex operator/ (double lhs, const Complex& rhs) {
        return Complex(lhs * rhs.real_ / (pow(rhs.real_, 2) + pow(rhs.imag_, 2)), lhs * rhs.imag_ / (pow(rhs.real_, 2) + pow(rhs.imag_, 2)) * -1);
    }

}  // namespace complex_numbers

File: solutions/python/complex-numbers/1/complex_numbers.py
import math

class ComplexNumber:
    def __init__(self, real, imaginary):
        self.real = real
        self.imaginary = imaginary

    def __eq__(self, other):
        return bool(self.real == other.real and self.imaginary == other.imaginary)

    def __add__(self, other):
        if not isinstance(other, ComplexNumber): other = ComplexNumber(other, 0)
        
        sum_real = self.real + other.real
        sum_imaginary = self.imaginary + other.imaginary

        return ComplexNumber(sum_real, sum_imaginary)

    def __radd__(self, other):
        if type(other) in (int, float):
            return self + other
    
    def __mul__(self, other):
        if not isinstance(other, ComplexNumber): other = ComplexNumber(other, 0)
        
        a, c = self.real, other.real
        b, d = self.imaginary, other.imaginary
            
        return ComplexNumber(a * c - b * d, b * c + a * d)

    def __rmul__(self, other):
        if type(other) in (int, float):
            return self * other

    def __sub__(self, other):
        if not isinstance(other, ComplexNumber): other = ComplexNumber(other, 0)
        print('sub!')

        dif_real = self.real - other.real
        dif_imaginary = self.imaginary - other.imaginary

        print(dif_real)
        print(dif_imaginary)
            
        return ComplexNumber(dif_real, dif_imaginary)

    def __rsub__(self, other):
        print('rsub!')
        if type(other) in (int, float):
            return ComplexNumber(other, 0) - self

    def __truediv__(self, other):
        if not isinstance(other, ComplexNumber): other = ComplexNumber(other, 0)

        a, c = self.real, other.real
        b, d = self.imaginary, other.imaginary
        
        return ComplexNumber((a * c + b * d)/(c**2 + d**2),(b * c - a * d)/(c**2 + d**2))

    def __rtruediv__(self, other):
        if type(other) in (int, float):
            return ComplexNumber(other, 0) / self

    def __abs__(self):
        return math.sqrt(self.real ** 2 + self.imaginary ** 2)

    def conjugate(self):
        return ComplexNumber(self.real, -self.imaginary)

    def exp(self):
        a, b = self.real, self.imaginary
        e = math.e

        return ComplexNumber(e ** a, 0) * (ComplexNumber(math.cos(b), 0) + ComplexNumber(0, math.sin(b)))